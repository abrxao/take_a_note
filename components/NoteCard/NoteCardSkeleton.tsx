"use client";
import React from "react";
import { Skeleton } from "../ui/skeleton";
import { Separator } from "@radix-ui/react-dropdown-menu";

export default function NoteCardSkeleton() {
  return (
    <div className="dark:bg-zinc-800 w-72 max-w-sm dark:border-zinc-700 dark:hover:bg-zinc-900 border border-zinc-300 px-4 py-2 rounded-lg">
      <Skeleton className="w-32 h-5" />
      <Separator className="my-2" />
      <div className="flex flex-col items-center gap-1">
        <Skeleton className="w-full h-5" />
        <Skeleton className="w-full h-5" />
      </div>
      <div className="flex justify-between items-center mt-4">
        <Skeleton className="w-full h-5" />
        <Skeleton className="w-full h-5" />
      </div>
    </div>
  );
}
