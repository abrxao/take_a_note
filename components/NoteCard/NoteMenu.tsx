"use client";
import { toast } from "sonner";
import { Edit2, X } from "lucide-react";
import useUserDataState from "@/stores/userDataStore";
import NoteProps from "@/types/note";
import { DrawingPinFilledIcon, DrawingPinIcon } from "@radix-ui/react-icons";
import useNoteDrawerState from "@/stores/noteModalStore";
import { ButtonHTMLAttributes } from "react";
import { twMerge } from "tailwind-merge";

interface NoteMenuProps {
  note: NoteProps;
}

function NoteMenuButton({
  children,
  className,
  ...props
}: ButtonHTMLAttributes<HTMLButtonElement>) {
  return (
    <button
      className={twMerge(
        "p-1 active:scale-8 dark:bg-zinc-700 hover:bg-zinc-200 dark:hover:bg-slate-400 rounded-md",
        className
      )}
      type="button"
      {...props}
    >
      {children}
    </button>
  );
}

export default function NoteMenu({ note }: NoteMenuProps) {
  const { removeNote, undoRemove, togglePinned } = useUserDataState();
  const { setEditNote, setIsDrawerNoteOpen } = useNoteDrawerState();
  return (
    <div className="flex gap-1">
      <NoteMenuButton
        className="hover:text-red-400"
        aria-label="Delete note"
        data-testid="note-menu-button-delete"
        onClick={() => {
          removeNote(note);
          toast("Note has been deleted", {
            description: `Click on 'undo' to get back "${note.title}" note`,
            action: {
              label: "undo",
              onClick: () => note.id && undoRemove(note.id),
            },
          });
        }}
      >
        <X size={14} />
      </NoteMenuButton>
      <NoteMenuButton
        data-value={note.favorite ? "checked" : ""}
        className=" dark:data-[value=checked]:bg-emerald-600 data-[value=checked]:bg-emerald-400"
        aria-label="Toggle favorite"
        data-testid="note-menu-button-favorite"
        onClick={() => {
          togglePinned(note);
        }}
      >
        {note?.favorite ? (
          <DrawingPinFilledIcon data-testid="favorite-icon" />
        ) : (
          <DrawingPinIcon data-testid="to-favorite-icon" />
        )}
      </NoteMenuButton>
      <NoteMenuButton
        data-testid="note-menu-button-edit"
        className=" hover:text-blue-600"
        aria-label="Edit note"
        type="button"
        onClick={() => {
          setIsDrawerNoteOpen(true);
          setEditNote(note);
        }}
      >
        <Edit2 size={14} />
      </NoteMenuButton>
    </div>
  );
}
