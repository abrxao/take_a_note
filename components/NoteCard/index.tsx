"use client";
import React, { ReactNode, useEffect, useMemo, useState } from "react";
import { Separator } from "../ui/separator";
import { Checkbox } from "../ui/checkbox";
import NoteMenu from "./NoteMenu";
import { CardDescription } from "../ui/card";
interface NoteCardProps {
  title: string;
  description: string;
}

export default function NoteCard(): ReactNode {
  const [isChecked, setIsChecked] = useState<boolean>(false);

  return (
    <div className="dark:bg-slate-500 border border-slate-300 px-4 py-2 rounded-lg">
      <h4 className="font-bold">Checkbox</h4>
      <Separator className="my-1" />
      <div className="flex items-center space-x-2">
        <Checkbox
          id="test"
          checked={isChecked}
          onClick={(e) => setIsChecked(!isChecked)}
        />
        <label htmlFor="test">Deploy your new project in one-click.</label>
      </div>
      <div className="flex justify-between items-center">
        <CardDescription className="dark:hover:text-slate-300 hover:text-slate-700">
          Editado
        </CardDescription>
        <NoteMenu />
      </div>
    </div>
  );
}
