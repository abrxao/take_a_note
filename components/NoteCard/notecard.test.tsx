import { act, fireEvent, render, screen } from "@testing-library/react";
import NoteCard from ".";
import NoteProps from "@/types/note";
import NoteCardSkeleton from "./NoteCardSkeleton";

describe("Note card tests", () => {
  const note: NoteProps = {
    title: "Nova nota",
    id: "3637",
    createDate: new Date("2024-01-15T11:37:33.749Z"),
    favorite: true,
    tags: ["The Love"],
  };
  const editedNote: NoteProps = {
    ...note,
    lastEditDate: new Date("2024-01-19T16:28:32.993Z"),
  };

  test("Render Note with only create date", async () => {
    render(<NoteCard {...note} />);
    const title = screen.getByText(note.title);
    expect(title).toBeInTheDocument();
  });
  test("Render Note with tags, create and last edited date", async () => {
    render(<NoteCard {...editedNote} />);
    const title = screen.getByText(editedNote.title);
    expect(title).toBeInTheDocument();
  });
  test("Simulation click on favorite", async () => {
    render(<NoteCard {...editedNote} />);
    const favoriteButton = screen.getByTestId("note-menu-button-favorite");
    await act(async () => {
      fireEvent.click(favoriteButton);
    });
    const iconFavorite = await screen.findByTestId("favorite-icon");
    expect(iconFavorite).toBeInTheDocument();
  });
  test("render skeleton", async () => {
    render(<NoteCardSkeleton />);
    const noteCardSkeleton = screen.getByTestId("note-card-skeleton");

    expect(noteCardSkeleton).toMatchSnapshot();
  });
});
