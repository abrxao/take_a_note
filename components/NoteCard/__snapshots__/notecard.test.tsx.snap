// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Note card tests render skeleton 1`] = `
<div
  class="dark:bg-zinc-800 w-72 max-w-sm dark:border-zinc-700 dark:hover:bg-zinc-900 border border-zinc-300 px-4 py-2 rounded-lg"
  data-testid="note-card-skeleton"
>
  <div
    class="animate-pulse rounded-md bg-slate-100 dark:bg-slate-800 w-32 h-5"
  />
  <div
    aria-orientation="horizontal"
    class="my-2"
    role="separator"
  />
  <div
    class="flex flex-col items-center gap-1"
  >
    <div
      class="animate-pulse rounded-md bg-slate-100 dark:bg-slate-800 w-full h-5"
    />
    <div
      class="animate-pulse rounded-md bg-slate-100 dark:bg-slate-800 w-full h-5"
    />
  </div>
  <div
    class="flex justify-between items-center mt-4"
  >
    <div
      class="animate-pulse rounded-md bg-slate-100 dark:bg-slate-800 w-full h-5"
    />
    <div
      class="animate-pulse rounded-md bg-slate-100 dark:bg-slate-800 w-full h-5"
    />
  </div>
</div>
`;
